src/high-level/vector.lisp

```diff
@@ -0,0 +1,167 @@
+                 (copy-seq (,storage-sym v)))
+           new-v))
+
+       (defmethod tref ((vector ,name) &rest pos)
+         (policy-cond:with-expectations (> speed safety)
+             ((assertion (valid-index-p pos (list (vector-size vector)))))
+           (aref (,storage-sym vector) (first pos))))
+
+       (defmethod (setf tref) (new-value (vector ,name) &rest pos)
+         (policy-cond:with-expectations (> speed safety)
+             ((assertion (valid-index-p pos (list (vector-size vector)))))
+           (setf (aref (,storage-sym vector) (first pos)) new-value))))))
+
+
+(defun pprint-vector (stream vector)
+  "Pretty-print a vector VECTOR to the stream STREAM."
+  (flet ((print-real (x)
+           (format stream "~6,3f" x))
+         (print-complex (z)
+           (format stream "~6,3f ~:[+~;-~]~6,3fj"
+                   (realpart z)
+                   (minusp (imagpart z))
+                   (abs (imagpart z))))
+         (print-int (x)
+           (format stream "~3d" x)))
+    (let* ((size (size vector))
+           (type (element-type vector))
+           (print-entry
+             (cond
+               ((subtypep type 'complex) #'print-complex)
+               ((subtypep type 'integer) #'print-int)
+               (t #'print-real))))
+      (pprint-logical-block (stream nil)
+        (print-unreadable-object (vector stream :type t)
+          (format stream "(~D):" size)
+          (dotimes (e size)
+            (pprint-newline :mandatory stream)
+            (funcall print-entry (tref vector e))))))))
+
+(set-pprint-dispatch 'vector 'pprint-vector)
+
+;;; Required abstract-tensor methods
+
+(defmethod order ((vector vector))
+  (declare (ignore vector))
+  1)
+
+(defmethod shape ((vector vector))
+  (list (vector-size vector)))
+
+(defmethod (setf shape) (new-value (vector vector))
+  (policy-cond:with-expectations (> speed safety)
+      ((type shape new-value)
+       (assertion (cl:= 1 (length new-value))))
+    (setf (vector-size vector) (first new-value))))
+
+(defgeneric dot (vector1 vector2)
+  (:documentation "Compute the dot product of two vectors")
+  (:method ((vector1 vector) (vector2 vector))
+    (policy-cond:with-expectations (> speed safety)
+        ((assertion (cl:= (size vector1) (size vector2))))
+      (loop :for i :below (size vector1)
+            :sum (* (tref vector1 i) (tref vector2 i))))))
+
+(defgeneric norm (vector &optional p)
+  (:documentation "Compute the p-norm of a vector.
+
+If P is not specified then the Euclidean norm is computed.
+Special values of P are: :INFINITY :INF :NEGATIVE-INFINITY :NEG-INF")
+  (:method ((vector vector) &optional (p 2))
+    (cond
+      ((or (eql p :infinity)
+           (eql p :inf))
+       (abs (reduce (lambda (x y) (max (abs x) (abs y))) (storage vector))))
+      ((or (eql p :negative-infinity)
+           (eql p :neg-inf))
+       (abs (reduce (lambda (x y) (min (abs x) (abs y))) (storage vector))))
+      (t
+       (assert (< 0 p)
```

Comment:
To be a real nerd about it, p must be greater than or equal to 1. For 0 < p < 1 the triangle inequality is not satisfied.

Reply:
☠️ 🔫 🤠  

@jmbr proves again he is the fastest reviewer in the west

Reply:
Ok nerd (fixed in 4a1b206)